### DISCUSSION TITLE
`[question]` How does the turtle keep moving during infinite recursion

### DISCUSSION THREAD

**Discussion by harold** [jrincayc/ucblogo-code#179] [General]
Hey! Found this repo today linked a couple hops from an older Papert/MIT page. Got the software running and made a few super-interesting (to me) experiments. Much appreciation for everyone's efforts here.

---

![image](https://github.com/jrincayc/ucblogo-code/assets/7443/637f02fc-3abe-422c-8ff2-4528820b0cca)

This code:

```
to t :i
right 80
forward i
wait 10
t i+2
end
```

Actually runs and draws things (as seen in the screenshot), a similar program, without `wait` ended up quickly filling the screen, due to the wrapping behavior, I guess.

This was surprising to me, because my intuition was that the recursion on `t` would happen 'immediately' and eventually overflow some stack, or at the very least hang and not give the program a chance to draw anything or otherwise continue to run.

---

My question:

What is the right way to think about how this is working?

Does `t` compile to (or somehow otherwise operate as) something that hands control back to the application outside the interpreter to let the turtle and GUI update? How often? Each call to `t`? Or periodically in time?

Or is it better thought of that the turtle is on another 'thread' (literal machine thread or otherwise) consuming commands on some queue being fed by the interpreter generally and `t` specifically?

Or is some other mental model better?

Similar question about `forever` in LOGO, how can that work?

Thanks in advance for any insights - having a lot of fun and learning a lot over here.

**Comment by brianharvey** [comment]
Yes, the display is a separate thread.

The reason your recursive procedure never runs out of stack is that the (only) recursive call is the last thing the procedure does, and Berkeley Logo does tail call elimination, which means that it essentially reuses the same stack frame instead of growing the stack.  If it were embedded recursion (with more code after the recursive call) then it would indeed eventually run out of stack space.

About the edge-of-screen behavior, you can tell Logo what you'd like it to do:
WRAP - go out one side and come back in the opposite side, i.e., treat the window as a torus.
WINDOW - the turtle moves on an infinite plane, only a little of which is visible.
FENCE - Logo gives an error message when the turtle runs into an edge.

**Reply by harold** [reply]
Outstanding - I got it now. Kind thanks for your time and consideration.

https://github.com/jrincayc/ucblogo-code/blob/2b161b7ddda246f6d3350b6b98e3200b48c2c9ab/graphics.c#L1710-L1732

Stay rad. :bow:
